### PR TITLE
Histogram: Add approx bucket count option

### PR DESCRIPTION
**Pull request by leeoniya**
this allows the Histogram panel and transformer to have higher data-adaptive resolution than the default 20 buckets. without this, the user would have to set a known bucket size to increase the resolution, which doesn't work well across different data sets.

this is also needed to auto-migrate `Graph (old)` in Histogram x-axis mode to the new Histogram panel:

image

TODO:

- [ ] Add Graph (old) migration code

### CODE REVIEW

**Comment by leeoniya** [comment]
so, it looks like the old histogram mode was width-adaptive. news to me!

<details><summary>old-graph-histogram-migration.json</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 922,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "aliasColors": {},
      "bars": true,
      "dashLength": 10,
      "dashes": false,
      "datasource": {
        "type": "grafana-testdata-datasource",
        "uid": "PD8C576611E62080A"
      },
      "editable": true,
      "error": false,
      "fieldConfig": {
        "defaults": {
          "unitScale": true
        },
        "overrides": []
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 7,
        "w": 16,
        "x": 0,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 1,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": false,
        "total": false,
        "values": false
      },
      "lines": false,
      "linewidth": 2,
      "links": [],
      "nullPointMode": "connected",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "10.4.0-pre",
      "pointradius": 5,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenario": "random_walk",
          "scenarioId": "random_walk",
          "target": ""
        }
      ],
      "thresholds": [],
      "timeFrom": "2s",
      "timeRegions": [],
      "title": "Millisecond res x-axis and tooltip",
      "tooltip": {
        "msResolution": false,
        "shared": false,
        "sort": 0,
        "value_type": "cumulative"
      },
      "type": "graph",
      "xaxis": {
        "mode": "histogram",
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "$$hashKey": "object:69",
          "format": "short",
          "logBase": 1,
          "show": true
        },
        {
          "$$hashKey": "object:70",
          "format": "short",
          "logBase": 1,
          "show": true
        }
      ],
      "yaxis": {
        "align": false
      }
    },
    {
      "datasource": {
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 80,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineWidth": 1
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "unitScale": true
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 16,
        "x": 0,
        "y": 7
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        }
      },
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 1,
          "refId": "A"
        }
      ],
      "title": "Panel Title",
      "type": "histogram"
    }
  ],
  "refresh": "",
  "schemaVersion": 39,
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "old graph histogram migration",
  "uid": "b872160a-2efc-4fe0-963e-b7ce0b841399",
  "version": 4,
  "weekStart": ""
}
```
</details>

https://github.com/grafana/grafana/assets/43234/93c323bb-e538-4f61-90a5-ccdd9e538dbd

File: packages/grafana-data/src/transformations/transformers/histogram.ts [hunk 1]
```diff
@@ -38,10 +38,13 @@ export const histogramBucketSizes = [
 ];
 /* eslint-enable */
 
+const DEFAULT_BUCKET_COUNT = 30;
```
Comment (baldm0mma) [inline packages/grafana-data/src/transformations/transformers/histogram.ts:41]: I think you mentioned the old default was 20? Is 30 enough to make a notable diff?

Comment (leeoniya) [reply packages/grafana-data/src/transformations/transformers/histogram.ts:41]: i think the increased resolution helps, and brings it closer to the old graph panel at 25%-50% panel widths, which is probably common (vs 100% histogram panel width)

File: public/app/features/transformers/editors/HistogramTransformerEditor.tsx [hunk 2]
```diff
@@ -105,6 +128,15 @@ export const HistogramTransformerEditor = ({
 
     return (
       <div>
+        <InlineFieldRow>
+          <InlineField
+            labelWidth={labelWidth}
+            label={histogramFieldInfo.bucketCount.name}
+            tooltip={histogramFieldInfo.bucketCount.description}
+          >
+            <NumberInput value={options.bucketCount} placeholder="20" onChange={onBucketCountChanged} min={0} />
```
Comment (baldm0mma) [inline public/app/features/transformers/editors/HistogramTransformerEditor.tsx:137]: Should the placeholder here be `30`?

File: public/app/features/transformers/editors/HistogramTransformerEditor.tsx [hunk 3]
```diff
@@ -138,6 +170,22 @@ export const HistogramTransformerEditor = ({
...
+          invalid={isInvalid.bucketCount}
+          error={'Value needs to be an integer or a variable'}
+          labelWidth={labelWidth}
+          label={histogramFieldInfo.bucketCount.name}
+          tooltip={histogramFieldInfo.bucketCount.description}
+        >
+          <SuggestionsInput
+            suggestions={variables}
+            value={options.bucketCount}
+            placeholder="20"
```
Comment (baldm0mma) [inline public/app/features/transformers/editors/HistogramTransformerEditor.tsx:184]: Same here as above?

Comment (leeoniya) [reply public/app/features/transformers/editors/HistogramTransformerEditor.tsx:184]: yeah, prolly

```
placeholder={`Default: ${defaultOptions.bucketCount}`}
```

File: public/app/features/transformers/editors/HistogramTransformerEditor.tsx (hunk 2 above)
Comment (leeoniya) [reply public/app/features/transformers/editors/HistogramTransformerEditor.tsx:137]: mhm

File: public/app/plugins/panel/histogram/module.tsx [hunk 4]
```diff
@@ -54,7 +63,9 @@ export const plugin = new PanelPlugin<Options, FieldConfig>(HistogramPanel)
     standardOptions: {
       [FieldConfigProperty.Color]: {
         settings: {
-          byValueSupport: true,
+          byValueSupport: false,
+          bySeriesSupport: true,
+          preferThresholdsMode: false,
```
Comment (leeoniya) [inline public/app/plugins/panel/histogram/module.tsx:68]: this is a drive-by fix, which makes sure the default color mode is classic palette, not thresholds.

**Comment by ryantxu** [comment]
> Add Graph (old) migration code

🤔 I think this may need to be a different (bigger) project.  The auto migrations now seems to be tied to the same panel type
https://github.com/grafana/grafana/blob/656fd4452e9c7d34927f7f3796a5aa55bc47f151/public/app/features/dashboard/state/DashboardModel.ts#L173

i'm sure there is a way, but seems bigger than this PR

**Comment by leeoniya** [comment]
> i'm sure there is a way, but seems bigger than this PR

we also need to consider the Graph -> BarChart case that will require appending the appropriate reduce transformation.

**Review by codeincarnate — APPROVED** [review]
Tested it out and seems to be working well! I can't comment on the broader implications and I'll leave that to Ryan and others, but in general looks good 😄

**Review by ryantxu — APPROVED** [review]
